### PR TITLE
refactor: use Postgres returning keyword to echo back removed IDs

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

In order to have a cache-friendly `DeleteByQuery`, the function should return the IDs it removed so the cache can update itself by removing the IDs that are not in the store anymore.
The previous approach of "fetch the IDs, then remove them" involves two roundtrips to the DB which is inefficient, and could actually miss some of the objects that should have been removed.

PR stack:
* #8947 
  * #8975
    * #8586 
      * #8827 
      * #8770  
      * #9216
      * #8682  
        * #8863  

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: pkg/search/postgres/common.go [hunk 1]
```diff
@@ -866,6 +882,61 @@ func RunDeleteRequestForSchema(ctx context.Context, schema *walker.Schema, q *v1
...
+	bufferToScanRowInto := make([]interface{}, numSelectFieldsForPrimaryKey)
+	if primaryKeysComposite {
+		var outputSlice []interface{}
+		bufferToScanRowInto[0] = &outputSlice
+	} else {
+		for i := 0; i < numPrimaryKeys; i++ {
+			bufferToScanRowInto[i] = pointers.String("")
+		}
+	}
+	returnedIDs := make([]string, 0)
```
Comment (connorgorman) [inline pkg/search/postgres/common.go:906]: var returnedIDs []string

**Review by connorgorman — COMMENTED** [review]
Generally,  this looks good. Two thoughts:
1. What are the performance implications of always returning from delete by query. I'd like to understand when we do a large delete by query how expensive it is (as the returned IDs are mostly only useful for the cases when we want to handle cache invalidation)
2. We should have some tests for this path so we can verify it's working correctly

**Review by janisz — COMMENTED** [review]
I miss the unit test of generated query. Integration test is fine but will be good to have something to validate how the generated query looks like

**Comment by rhybrillou** [comment]
> I miss the unit test of generated query. Integration test is fine but will be good to have something to validate how the generated query looks like

There are two new query generator tests in `common_test.go`: one for `DELETE` (L 408 to 561) and one for `DELETERETURNINGIDS` (L 563 to 728). Are you requesting additional tests on top of these ?

**Review by connorgorman — APPROVED** [review]

File: pkg/search/postgres/common.go [hunk 2]
```diff
@@ -194,6 +196,8 @@ func (q *query) getPortionBeforeFromClause() string {
 	switch q.QueryType {
 	case DELETE:
 		return "delete"
+	case DELETERETURNINGIDS:
+		return "delete"
```
Comment (dhaus67) [inline pkg/search/postgres/common.go:200]: super nit:
```suggestion
	case DELETERETURNINGIDS, DELETE:
		return "delete"
```

**Review by dhaus67 — APPROVED** [review]

**Pull request merged**